drgn_tools/locking.py

```diff
@@ -519,3 +570,92 @@ def get_rwsem_info(rwsem: Object, callstack: int = 0) -> None:
         print("There are no waiters")
     else:
         get_rwsem_waiters_info(rwsem, callstack)
+
+
+def is_task_blocked_on_lock(
+    pid: IntegerLike, lock_type: str, lock: Object
+) -> bool:
+    """
+    Check if a task is blocked on a given lock or not
+    :param pid: PID of task
+    :param var_name: variable name (sem, or mutex)
+    :param lock: ``struct mutex *`` or ``struct semaphore *`` or ``struct rw_semaphore *``
+    :returns: True if task is blocked on given lock, False otherwise.
+    """
+
+    try:
+        if lock_type == "semaphore" or lock_type == "rw_semaphore":
+            return pid in [
+                waiter.pid.value_()
+                for waiter in for_each_rwsem_waiter_careful(lock.prog_, lock)
+            ]
+        elif lock_type == "mutex":
+            return pid in [
+                waiter.pid.value_()
+                for waiter in for_each_mutex_waiter_careful(lock.prog_, lock)
+            ]
+        else:
+            return False
+    except (FaultError, ValidationError):
+        return False
+
+
+def get_lock_from_frame(
+    prog: Program, task: Object, frame: StackFrame, kind: str, var: str
+) -> Optional[Object]:
+    """
+    Given a stack frame, try to get the relevant lock out of it.
+
+    :param task: the ``struct task_struct *`` of the task
+    :param frame: the stack frame in question
+    :param kind: the kind of lock (mutex, rw_semaphore, semaphore)
+    :param var: the variable name within the stack frame
+    :returns: an object of the appropriate type for the lock kind. If it could
+      not be found, returns None
+    """
+    # Try to use DWARF CFI to get the variable value. This is the most
+    # straightforward method to get the lock variable.
+    try:
+        lock = frame[var]
+        if not lock.absent_:
+            return lock
+    except (drgn.ObjectAbsentError, KeyError):
+        pass
+
+    # Fall back to a brute force method which works shockingly well. Simply scan
+    # from the top of the stack down to the frame of the lock. If the address is
+    # a kernel address, we'll test it (carefully) to see whether it's a valid
+    # lock, and if so, we'll return it.
+    #
+    # First, we need to get the start and end address to search. This is
+    # arch-specific.
+    if prog.platform.arch == Architecture.X86_64:
+        candidates = range(task.thread.sp, frame.sp, 8)
```

Comment:
`frame.sp` is the bottom of the stack?

Reply:
Yeah, it's the bottom of this stack frame. This reliably works well though! I think it's because x86 always spills the register beneath the stack frame anyway. With aarch64, it seems that the register gets stored within the stack frame.

I could try to get the next stack frame up, but this function doesn't actually accept a stack trace... just the single stack frame. I could try to use the workaround with the frame pointer that aarch64 does. In practice, we always find the lock before we get to the upper end up the range anyway, so it wouldn't be much of an issue.

Reply:
so `task.thread.sp` is the bottom of the stack, then the scan happened to the parents stack frame of current one, could there be a case that lock address is stored in the children stack frame of current one?

Reply:
I guess I'm not sure what you mean. The `task.thread.sp` is the "bottom" of the stack, that is, it is where the stack pointer was stored just prior to context switching away. Then, `frame.sp` is the stack pointer for the current function (e.g. `mutex_lock()`). So, the stack looks something like this:

```
HIGHER ADDRESSES / OLDER FRAMES HERE
[parent caller frame]
...
[mutex_lock() frame]
0xDEADBEEF
0xF000B000
... vars ...                                       <---- frame.sp
[schedule_preempt_disabled() frame]
...
[schedule() frame]
... vars ...
... bottom of the stack ...                     <--- task.thread.sp
LOWER ADDRESSES / NEWER FRAMES HERE
```

So we start at `task.thread.sp` and increment the address until we reach `frame.sp`. Since `frame.sp` is technically pointing at the bottom of the stack frame for `mutex_lock()`, it's possible that the lock could be somewhere in that frame and we'd miss it. However in practice that just doesn't happen, the lock has always been beneath the frame.sp value.

Reply:
My fault, thought `task.thread.sp` is the other end of the stack. 
I'm OK with current approach, we can always fix it if lock is stored in current stack frame in future kernels.